### PR TITLE
feat(pr): #266 env var + GraphQL rate-limit polish + 405 sha-gated merge retry

### LINKED ISSUE #266: feat: expose SHIPYARD_PR_RUNNING=1 env var to child processes; back off on GraphQL rate-limit
### 1. `SHIPYARD_PR_RUNNING=1` env var (signal)

Please set `SHIPYARD_PR_RUNNING=1` (or similar) in the environment of every subprocess `shipyard pr` spawns — particularly `git push` and `gh api`. This lets downstream pre-push hooks reliably detect that a push is happening under shipyard supervision (vs a direct `git push` that bypasses local validation).

**Use case**: pulp's pre-push hook now logs every push to a drift log so we can audit when PRs went up via shipyard (full local validation) vs direct push (skips skill-sync / version-bump / diff-coverage / SSH-host validation, turns CI into the discovery channel). Without a clean signal from shipyard, we have to use a user-set fallback (`PULP_VIA_SHIPYARD=1`) which is error-prone — easy to forget when running shipyard.

Reference: pulp PR #1406 (https://github.com/danielraffel/pulp/pull/1406) is the consumer-side hook that would benefit from this.

### 2. Back off cleanly on GraphQL rate-limit (bug)

Observed 2026-05-04 (shipyard v0.46.0): `shipyard pr` retried PR creation 3× when GitHub returned `API rate limit already exceeded for user ID 25807` (GraphQL quota). Each retry burned more GraphQL quota and pushed reset further out, then hard-failed with an unhandled `GhError` traceback dumped to stdout.

Expected behavior: detect rate-limit error code (`graphql_rate_limit` / HTTP 403 with secondary-rate-limit headers), parse the `X-RateLimit-Reset` header or use `gh api rate_limit`, and either:
- Wait for reset (clear message: "GraphQL exhausted; reset at HH:MM:SS UTC; waiting")
- Fall back to REST `POST /repos/{owner}/{repo}/pulls` for PR creation (the merge step still needs GraphQL but PR creation has a REST endpoint)
- Exit cleanly with a clear error and recovery suggestion, NO traceback

```
File "shipyard/ship/pr.py", line 67, in create_pr
File "shipyard/ship/pr.py", line 193, in _run_gh
shipyard.ship.pr.GhError: GraphQL: API rate limit already exceeded for user ID 25807.
[PYI-3700:ERROR] Failed to execute script 'cli' due to unhandled exception!
```

The naked traceback + retry burn was disorienting in a long PR session.

### Why these matter together

When GraphQL is the bottleneck (which seems to happen reliably in any session that opens 3-4+ PRs), `shipyard pr` becomes unusable AND the user has no clean fallback signal — they have to remember to set `PULP_VIA_SHIPYARD=1` on every direct push to maintain audit trail. Both fixes together make `shipyard pr` more graceful under quota pressure.

### Environment
- shipyard 0.46.0 (pinned via pulp/tools/shipyard.toml)
- gh CLI 2.x
- macOS 26.x ARM64
- Heavy-user pattern: 5-8 PRs in flight, mix of macOS-local + SSH Ubuntu/Windows lanes

### PR DESCRIPTION
**Pull request by danielraffel**
Closes #266.

Implements both asks from #266 (and the 2026-05-16 405-retry follow-up comment) in two commits on the same branch — separate for clean revert if either slice misbehaves.

## Commit 1 — `feat(pr): set SHIPYARD_PR_RUNNING=1 on supervised subprocesses`

Marks every `git` / `gh` child spawned by Shipyard's supervised flows (`pr`, `ship`, `auto-merge`, `overflow`, `wait`) with `SHIPYARD_PR_RUNNING=1`. Downstream consumers — notably the Pulp pre-push hook in [danielraffel/pulp#1406](https://github.com/danielraffel/pulp/pull/1406), which expects this exact env var name — can now distinguish Shipyard-orchestrated pushes from raw `git push` without requiring users to set a project-local fallback marker manually.

**Scope discipline:** In scope: `pr.rs`, `branch.rs`, `wait_transport.rs`, `app/pr_cmd.rs`, `app/ship_cmd.rs`, `app/auto_merge_cmd.rs`, `preflight.rs`. Out of scope: diagnostic / maintenance subcommands (`doctor`, `pin`, `runner`, `cleanup`, `cloud`, `governance`, `release_bot`, `reconcile`) — they're not supervised pushes per the audit-log use case.

**Implementation:** New `src/supervised.rs` module with three `pub(crate)` helpers (`supervised`, `gh_supervised`, `git_supervised`) + constants for the env-var name and value (single source of truth). `pr.rs`'s existing `gh()` helper delegates to `gh_supervised`. `branch.rs`'s test-injection point also wrapped so the marker survives shim injection. 22 direct `Command::new` sites replaced via mechanical sed.

## Commit 2 — `fix(pr,merge): surface GraphQL rate-limit fallback + 405 sha-gated merge retry`

Two polish items after v0.56.2 already shipped the REST-fallback skeleton:

**Rate-limit fallback notice.** Prior to this commit, all five REST fallbacks happened silently — users couldn't tell GraphQL had bailed. Adds `pr::report_rate_limit_fallback(operation, cwd)`:

```
shipyard: GraphQL rate limit hit for gh pr merge; reset at HH:MM:SS UTC. Falling back to REST.
```

Reset time comes from a best-effort `gh api rate_limit --jq .resources.graphql.reset` probe. `rate_limit` is in its own quota bucket (does NOT consume GraphQL budget), so the probe is safe from inside a recovery path. Probe failures silently omit the reset time. Wired into all 5 dispatch sites.

**Race-protected REST merge + 405 retry.** Per the 2026-05-16 comment on #266 ("on `405 Base branch was modified`, refresh and retry once only if the PR head SHA and green status still match"). Sharpened in Codex peer review: pass `sha=<head_sha>` on the REST merge PUT so GitHub enforces the race-guard *server-side* rather than client-side only.

`merge_pr_rest` now:
1. Fetches `head.ref` AND `head.sha` upfront (supersedes `pr_head_branch_rest` which only returned the ref).
2. Passes `head_sha` as `-f sha=<oid>` on the PUT → GitHub itself enforces "merge what I saw."
3. On HTTP 405 + body matching "Base branch was modified": re-fetches head info; if the head SHA is unchanged (i.e., modification was purely on the base branch — typical when a sibling PR lands during the merge attempt), retries the PUT exactly once.
4. If head SHA *did* change, refuses to retry and surfaces a clear `head moved from <abc1234> to <def5678>` error — operator re-evaluates because the prior green evidence may no longer apply.

## Tests

- `src/supervised.rs`: 5 unit tests (helper API contract + end-to-end `sh -c 'echo $SHIPYARD_PR_RUNNING'` round-trip).
- `src/app/auto_merge_cmd.rs::tests`: 6 new cases — 4 for `is_base_modified_405` detector edge cases (canonical body, case-insensitivity, rejecting unrelated 405s, rejecting the phrase without the 405 status code), 2 for `short_sha` helper.
- Full lib suite: **667/667 passing** (was 661 before; +6 new cases. No regressions.)

## What was already done in v0.56.2 (verified)

Confirmed by code reading + Codex peer review: the v0.46.0 behavior described in the issue body (3× retry burn + naked traceback) is gone. v0.56.2 already had REST fallback for PR list, create, view, snapshot, AND merge (`merge_pr_rest` at `auto_merge_cmd.rs:311`, complete with an inline comment citing the #266 scenario). This PR only adds the missing user-facing notice + the 405 polish.

## Alignment with Codex review

I ran the design past `codex exec` before implementing. Codex confirmed:
- v0.56.2-already-done finding is correct.
- Keep env-var and rate-limit-polish as separate commits (clean revert path > single-CHANGELOG-line).
- Use `pub(crate)` free helpers, not extension trait / typed wrapper.
- Scope env var to orchestrated flows (`pr` / `ship` / `auto-merge` / `overflow`), not literally only `pr`.
- Sharpened my client-side head_sha gate → pass `sha=` on the PUT for server-side enforcement.
- Don't refactor the 5 fallback dispatches into a higher-order helper now (keeps the regression surface tight).
- Verified pulp PR #1406 expects `SHIPYARD_PR_RUNNING` exactly.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/supervised.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+    #[test]
+    fn marker_propagates_to_child_process() {
+        // End-to-end: spawn a real subprocess and assert it sees the env.
+        // Uses `sh -c 'echo $SHIPYARD_PR_RUNNING'` because that is portable
+        // across the macOS / Linux runners shipyard supports.
+        let output = supervised(Command::new("sh"))
+            .args(["-c", "echo $SHIPYARD_PR_RUNNING"])
+            .output()
+            .expect("spawn sh");
```
[the remark this answers is not in the document]
Comment (danielraffel) [reply src/supervised.rs:121]: Fixed in `402e9dc`. Split the test into `marker_propagates_to_child_process_unix` (`#[cfg(unix)]`, uses `sh -c`) and `marker_propagates_to_child_process_windows` (`#[cfg(windows)]`, uses `cmd /C echo %SHIPYARD_PR_RUNNING%`). Each platform's CI runner exercises the contract with a command it actually has. Same commit also adds skill-sync notes to `skills/ci/SKILL.md` and `skills/shipyard/SKILL.md` plus a 0.56.2 → 0.57.0 minor bump for the new env-var + rate-limit-notice contracts.
